packages/batch/src/BasePartialProcessor.ts

```diff
@@ -66,13 +74,45 @@ abstract class BasePartialProcessor {
    * Call instance's handler for each record
    * @returns List of processed records
    */
-  public async process(): Promise<(SuccessResponse | FailureResponse)[]> {
+  public process(): (SuccessResponse | FailureResponse)[] {
+    /**
+     * If this is an async processor, user should have called processAsync instead,
+     * so we call the method early to throw the error early thus failing fast.
+     */
+    if (this.constructor.name === 'AsyncBatchProcessor') {
+      this.processRecord(this.records[0]);
+    }
     this.prepare();
 
     const processedRecords: (SuccessResponse | FailureResponse)[] = [];
     for (const record of this.records) {
-      processedRecords.push(await this.processRecord(record));
+      processedRecords.push(this.processRecord(record));
+    }
+
+    this.clean();
+
+    return processedRecords;
+  }
+
+  /**
+   * Call instance's handler for each record
+   * @returns List of processed records
+   */
+  public async processAsync(): Promise<(SuccessResponse | FailureResponse)[]> {
+    /**
+     * If this is an sync processor, user should have called process instead,
+     * so we call the method early to throw the error early thus failing fast.
+     */
+    if (this.constructor.name === 'BatchProcessor') {
+      await this.asyncProcessRecord(this.records[0]);
     }
+    this.prepare();
+
+    const processingPromises: Promise<SuccessResponse | FailureResponse>[] =
```

Comment:
I might be lacking understanding on how to handle asynchronous processes here, so I'm glad to learn more. 
Is there a reason not to do `push(await this.asyncProcessRecord)` to push a `SuccessResponse | FailureResponse` (matching the syntax of the synchronous processing) instead of doing a `Promise.all`?

Reply:
Ah, the comment didn't leave particularly where I intended it to be -- I'm referring to lines 111-115 here.

Reply:
No problem, let me explain: if we did `push(await this.asyncProcessRecord)` then the result of each `this.asyncProcessRecord` would be `await`ed before being added to the array. This would allow the customer to pass an async handler but the records would still be processed sequentially.

Using this method we are instead iterating the records and adding the unresolved promise to the array (notice the lack of `await`). After all the promises are in the array we call `Promise.all` which starts starts all the promises and resolves itself when all the underlying promises have either resolved or rejected.

This allows us to essentially process the whole batch concurrently (aka in parallel) rather than sequentially. I implemented the method this way to match the use case of the one [found in Python](https://docs.powertools.aws.dev/lambda/python/latest/utilities/batch/#processing-messages-asynchronously).

This detail (concurrent vs parallel) is also the reason why there's no `AsybcSQSFifoProcessor`, with FIFO queues we must respect the order of the records, so we can only process them sequentially.